### PR TITLE
Pg15 support

### PR DESCRIPTION
**Pull request by naisila**
DESCRIPTION: Adds support for PostgreSQL 15beta3

Helpful small PRs to main branch:

- https://github.com/citusdata/citus/pull/6211
- https://github.com/citusdata/citus/pull/6192 / https://github.com/citusdata/citus/pull/6184
- https://github.com/citusdata/citus/pull/6183
- https://github.com/citusdata/citus/pull/6186
- https://github.com/citusdata/citus/pull/6180
- https://github.com/citusdata/citus/pull/6143
- https://github.com/citusdata/citus/pull/6147
- https://github.com/citusdata/citus/pull/6145
- https://github.com/citusdata/citus/pull/6144
- https://github.com/citusdata/citus/pull/6142
- https://github.com/citusdata/citus/pull/6136
- https://github.com/citusdata/citus/pull/6118
- https://github.com/citusdata/citus/pull/6139
- https://github.com/citusdata/citus/pull/6110

### CODE REVIEW

File: src/backend/distributed/commands/collation.c [hunk 1]
```diff
@@ -63,24 +63,62 @@ CreateCollationDDLInternal(Oid collationId, Oid *collowner, char **quotedCollati
...
+	{
+		collcollate = TextDatumGetCString(datum);
+	}
+	else
+	{
+		collcollate = NULL;
+	}
+
 	datum = SysCacheGetAttr(COLLOID, heapTuple, Anum_pg_collation_collctype, &isnull);
-	Assert(!isnull);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/collation.c:73]: should we keep the `Assert(!isnull)` on PG 14/13 code?

File: src/test/regress/expected/distributed_collations.out [hunk 2]
```diff
@@ -1,8 +1,42 @@
 SET citus.next_shard_id TO 20050000;
+SHOW server_version \gset
+SELECT substring(:'server_version', '\d+')::int > 14 AS server_version_above_fourteen
+\gset
 CREATE USER collationuser;
 CREATE SCHEMA collation_tests AUTHORIZATION collationuser;
 CREATE SCHEMA collation_tests2 AUTHORIZATION collationuser;
 SET search_path to collation_tests;
+\if :server_version_above_fourteen
+-- fail, needs "locale"
```
Comment (onderkalaci) [inline src/test/regress/expected/distributed_collations.out:10]: can we have these new tests on pg15.sql such that we avoid alternative output file?

File: src/backend/distributed/commands/collation.c [hunk 3]
```diff
@@ -63,24 +63,62 @@ CreateCollationDDLInternal(Oid collationId, Oid *collowner, char **quotedCollati
...
+	}
+	else
+	{
+		collctype = NULL;
+	}
+
+	datum = SysCacheGetAttr(COLLOID, heapTuple, Anum_pg_collation_colliculocale, &isnull);
+	if (!isnull)
+	{
+		colliculocale = TextDatumGetCString(datum);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/collation.c:105]: Don't we need `pstrdup` for PG 15 part?

File: src/backend/distributed/commands/collation.c (hunk 1 above)
Comment (naisila) [reply src/backend/distributed/commands/collation.c:73]: No. Possibillity of entries being null is something introduced in PG15.

File: src/backend/distributed/commands/collation.c (hunk 3 above)
Comment (naisila) [reply src/backend/distributed/commands/collation.c:105]: No, because in PG15, collcollate and collctype are of type text, not name.

File: src/backend/distributed/commands/trigger.c [hunk 4]
```diff
@@ -183,7 +198,22 @@ GetExplicitTriggerIdList(Oid relationId)
...
+#if PG_VERSION_NUM >= PG_VERSION_15
+
+		/*
+		 * Pre PG15, tgisinternal is true for a "child" trigger on a partition
+		 * cloned from the trigger on the parent.
+		 * In PG15, tgisinternal is false in that case. However, we don't want to
+		 * create this trigger on the partition since it will create a conflict
+		 * when we try to attach the partition to the parent table:
+		 * ERROR: trigger "..." for relation "{partition_name}" already exists
+		 */
```
Comment (onderkalaci) [inline src/backend/distributed/commands/trigger.c:210]: I'm mostly wondering about existing such triggers on upgrades to PG 15.

Say, on PG 14 you have partitioned tables with triggers defined (and also maybe foreign keys as they create internal triggers -- as mentioned on f4566345cf40b068368cb5617e61318da60676ec).

And, you upgrade to PG 15. Can you still do:

- Create a new partition
- Rename / DROP existing triggers  before creating a new partition
- Rename / DROP existing triggers after creating a new partition

I think it'd be great to have such an upgrade test on our TODO list. We don't have to prioritize adding such test now, but seems could have some edge cases here.

Comment (naisila) [reply src/backend/distributed/commands/trigger.c:210]: > Say, on PG 14 you have partitioned tables with triggers defined (and also maybe foreign keys as they create internal triggers -- as mentioned on f4566345cf40b068368cb5617e61318da60676ec).
And, you upgrade to PG 15. Can you still do:
> * Create a new partition
> * Rename / DROP existing triggers before creating a new partition
> * Rename / DROP existing triggers after creating a new partitionThis change doesn't change the behavior of those actions

This change shouldn't change the behavior of those actions. If they were working before, they should work now. If they weren't working before, they shouldn't work now.
I mean, we were skipping triggers for partitions before because they were marked as internal. This change simply and only makes sure we still skip triggers for partitions.

Comment (onderkalaci) [reply src/backend/distributed/commands/trigger.c:210]: makes sense, so should be trivial to add the tests before/after we merge

File: src/test/regress/sql/multi_insert_select_conflict.sql [hunk 5]
```diff
@@ -1,3 +1,11 @@
+--
+-- MULTI_INSERT_SELECT_CONFLICT
+--
+-- This test file has an alternative output because of the change in the
```
Comment (JelteF) [inline src/test/regress/sql/multi_insert_select_conflict.sql:4]: In addition to this comment you should include a version check at the top, like we did in the past here:
https://github.com/citusdata/citus/commit/f9b02946b125483e1439d374129541c893592db4#diff-bddd1d49aa3304cb7531b4672f2ef9a6a502a8f75f03e1b03e6e4cfa4e191ee3L5-L8

Otherwise when an error occurs due to an actual bug it might compare against the wrong output file and then the diff will be much larger than intended.

File: src/test/regress/sql/citus_local_tables_queries.sql [hunk 6]
```diff
@@ -1,3 +1,15 @@
...
+-- CITUS_LOCAL_TABLES_QUERIES
+--
+-- Test queries on a distributed table with shards on the coordinator
+--
+-- This test file has an alternative output because of the change in the
+-- display of SQL-standard function's arguments in INSERT/SELECT in PG15.
+-- The alternative output can be deleted when we drop support for PG14
+--
+SHOW server_version \gset
+SELECT substring(:'server_version', '\d+')::int > 10 AS server_version_above_fourteen;
```
Comment (hanefi) [inline src/test/regress/sql/citus_local_tables_queries.sql:11]:
```suggestion
SELECT substring(:'server_version', '\d+')::int > 14 AS server_version_above_fourteen;
```

I guess there is a copy paste error here. I wonder how that does not result in test failures.

Please go over similar changes in other files as well.

Comment (naisila) [reply src/test/regress/sql/citus_local_tables_queries.sql:11]: Hi Hanefi, the PR is not ready for review

**Comment by naisila** [comment]
Note: this PR had earlier done very similar changes to the latest commits in the main branch. I have dropped those commits and force pushed since they were now redundant.
https://github.com/citusdata/citus/commit/e33ba7da9e750824c9bd8dc3b3c137af3143e99b
https://github.com/citusdata/citus/commit/616b1758c2b07154e74a6ae51fddf375f62d0172

File: src/test/recovery/Makefile [hunk 7]
```diff
@@ -12,6 +12,22 @@ subdir = src/test/recovery
...
 
+# In PG15, Perl test modules have been moved to a new namespace
+# new() and get_new_node() methods have been unified to 1 method: new()
+# Relevant PG commits 201a76183e2056c2217129e12d68c25ec9c559c8
+#                     b3b4d8e68ae83f432f43f035c7eb481ef93e1583
+pg_version = $(shell $(PG_CONFIG) --version 2>/dev/null)
+pg_whole_version = $(shell echo "$(pg_version)"| sed -e 's/^PostgreSQL \([0-9]*\)\(\.[0-9]*\)\{0,1\}\(.*\)/\1\2/')
+pg_major_version = $(shell echo "$(pg_whole_version)"| sed -e 's/^\([0-9]\{2\}\)\(.*\)/\1/')
+
+ifeq ($(pg_major_version),13)
```
Comment (onderkalaci) [inline src/test/recovery/Makefile:23]: Hmm, this works just because we have a single test file. If we had more than 1, we'd not be running this. (same for tap tests)

Can you see another way for doing this such that we wouldn't prevent future test additions?

File: src/test/regress/expected/citus_local_tables_queries.out [hunk 8]
```diff
@@ -1,3 +1,17 @@
+--
+-- CITUS_LOCAL_TABLES_QUERIES
+--
+-- This test file has an alternative output because of the change in the
+-- display of SQL-standard function's arguments in INSERT/SELECT in PG15.
+-- The alternative output can be deleted when we drop support for PG14
+--
+SHOW server_version \gset
+SELECT substring(:'server_version', '\d+')::int > 14 AS server_version_above_fourteen;
```
[the remark this answers is not in the document]
Comment (naisila) [reply src/test/regress/expected/citus_local_tables_queries.out:9]: Good idea. I chose this way to be in line of what we have chosen before. We have `server_version_above_thirteen` in the code, and when I dropped PG12 support, we had a lot of `server_version_above_twelve`. I will change it here, @jeff-davis already has chosen `server_version_ge_15` in the columnar pg15 tests that he has added, so I will go with that.
I will also open a PR to main to rename `server_version_above_thirteen` to `server_version_ge_14`.

File: src/backend/distributed/commands/trigger.c [hunk 9]
```diff
@@ -183,7 +198,22 @@ GetExplicitTriggerIdList(Oid relationId)
...
+		bool isChildTrigger = partitionTable && TriggerExists(parentRelationId, NameStr(
+																  triggerForm->tgname));
```
[the remark this answers is not in the document]
Comment (naisila) [reply src/backend/distributed/commands/trigger.c:212]: Awesome! I hadn't noticed there was such an entry in `pg_trigger` catalog. In that case, I also don't need to do previous partition/parent table checks.

File: src/test/recovery/Makefile (hunk 7 above)
Comment (naisila) [reply src/test/recovery/Makefile:23]: Ideally, I'd want to add the version check inside the test, but I didn't find a way of doing that. I could create two different folders. Any new test we add would need to include the Postgres node differently for different PG versions anyway.

File: src/test/columnar_freezing/Makefile [hunk 10]
```diff
@@ -10,6 +10,22 @@ subdir = src/test/columnar_freezing
...
 
+# In PG15, Perl test modules have been moved to a new namespace
+# new() and get_new_node() methods have been unified to 1 method: new()
+# Relevant PG commits 201a76183e2056c2217129e12d68c25ec9c559c8
+#                     b3b4d8e68ae83f432f43f035c7eb481ef93e1583
+pg_version = $(shell $(PG_CONFIG) --version 2>/dev/null)
+pg_whole_version = $(shell echo "$(pg_version)"| sed -e 's/^PostgreSQL \([0-9]*\)\(\.[0-9]*\)\{0,1\}\(.*\)/\1\2/')
+pg_major_version = $(shell echo "$(pg_whole_version)"| sed -e 's/^\([0-9]\{2\}\)\(.*\)/\1/')
+
+ifeq ($(pg_major_version),13)
```
Comment (onderkalaci) [inline src/test/columnar_freezing/Makefile:21]: maybe add few lines more comment:
```
# for now, we only have a single test file
# due to the above explanation, we ended up separating the test paths for
# different versions. If you need to add new test files, be careful to add both versions
```

File: src/test/regress/expected/multi_select_distinct.out [hunk 11]
```diff
@@ -3,6 +3,16 @@
 --
 -- Tests select distinct, and select distinct on features.
 --
+-- This test file has an alternative output because some HashAggregate
```
Comment (onderkalaci) [inline src/test/regress/expected/multi_select_distinct.out:6]: Are there any coordinator plan changes here? Can't we use coordinator_plan() trick to avoid additional test output file? We really don't care much what the worker queries generate in these tests

Comment (naisila) [reply src/test/regress/expected/multi_select_distinct.out:6]: Makes sense, thanks

**Review by onderkalaci — APPROVED** [review]
Great to see this land! 

There are few more things to do, but it is also more than ready to get into the development branches.
